include/boost/fusion/container/vector/vector.hpp

```diff
@@ -9,12 +9,310 @@
+                return *this;
+            }
+
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            T      & get()       { return elem; }
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            T const& get() const { return elem; }
+
+            T elem;
+        };
+
+        template <typename I, typename ...T>
+        struct vector_data;
+
+        template <std::size_t ...I, typename ...T>
+        struct vector_data<detail::index_sequence<I...>, T...>
+            : store<I, T>...
+            , sequence_base<vector_data<detail::index_sequence<I...>, T...> >
+        {
+            typedef vector_tag                  fusion_tag;
+            typedef fusion_sequence_tag         tag; // this gets picked up by MPL
+            typedef mpl::false_                 is_view;
+            typedef random_access_traversal_tag category;
+            typedef mpl::int_<sizeof...(T)>     size;
+            typedef vector<T...>                type_sequence;
+
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            vector_data()
+            {}
+
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            vector_data(copy_or_move, vector_data const& rhs)
+                : store<I, T>(static_cast<store<I, T> const&>(rhs))...
+            {}
+
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            vector_data(copy_or_move, vector_data&& rhs)
+                : store<I, T>(std::forward<store<I, T> >(rhs))...
+            {}
+
+            template <typename Sequence>
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            explicit
+            vector_data(from_sequence<detail::index_sequence<I...> >, Sequence&& rhs)
+                : store<I, T>(forward_at_c<I>(rhs))...
+            {}
+
+            template <typename ...U>
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            explicit
+            vector_data(each_elem, U&&... var)
+                : store<I, T>(std::forward<U>(var))...
+            {}
+
+            template <typename Sequence>
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            void
+            assign(Sequence&&, detail::index_sequence<>) {}
+
+            template <typename Sequence, std::size_t N, std::size_t ...M>
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            void
+            assign(Sequence&& seq, detail::index_sequence<N, M...>)
+            {
+                at_impl(mpl::int_<N>()) = vector_detail::forward_at_c<N>(seq);
+                assign(std::forward<Sequence>(seq), detail::index_sequence<M...>());
+            }
+
+            typedef extension::value_at_impl<vector_tag> value_at_impl;
+
+            template <typename J>
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            typename value_at_impl::template apply<vector_data, J>::type&
+            at_impl(J)
+            {
+                typedef typename value_at_impl::template apply<vector_data, J>::type U;
```

Comment:
I think it would be easier to have at_impl be a static function, something like:

```c++
template<std::size_t index, typename J>
static BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED const J& at_impl(const store<index, J>& value)
{
    return value.get();
}
```

That way the compiler does a search against the inheritance tree of `vector_data`, and deduces `J` from that search. I just tested this:
```c++
#include <iostream>
#include <utility>

template<std::size_t index, typename T>
struct store
{
        store(const T& t) : value(t) {}
        T value;
};

template<typename Indices, typename... T>
struct vector_base;

template<std::size_t ...Indices, typename ...T>
struct vector_base<std::integer_sequence<std::size_t, Indices...>, T...>
        : store<Indices, T>...
{
        vector_base(const T&... t)
                : store<Indices, T>(t)...
        {
        }
};

template<typename... T>
struct vector : vector_base<std::make_index_sequence<sizeof...(T)>, T...>
{
        using base = vector_base<std::make_index_sequence<sizeof...(T)>, T...>;

        vector(const T&... t)
                : base(t...)
        {
        }
};

template<std::size_t index, typename T>
const T& get(const store<index, T>& value)
{
        return value.value;
}


int main()
{
        using my_vector = vector<int, float, int, float>;

        my_vector test(10, 0.5f, 100, 1.5f);
        std::cout << get<2>(test) << std::endl;
        return 0;
}
```
in clang 3.5 and it works. Hopefully MSVC doesn't strike again here, I _think_ this would be a lot quicker at compile time. It would remove the linear instantation sequence (via `value_at_impl`), and that metafunction could be updated to use this static function + decltype.